homeassistant/components/envisalink.py

```diff
@@ -77,119 +84,111 @@
-                                          _pass,
-                                          _zone_dump,
-                                          _keep_alive,
-                                          hass.loop)
 
+    conf = config.get(DOMAIN)
+
+    host = conf.get(CONF_EVL_HOST)
+    port = conf.get(CONF_EVL_PORT)
+    code = conf.get(CONF_CODE)
+    panel_type = conf.get(CONF_PANEL_TYPE)
+    panic_type = conf.get(CONF_PANIC)
+    version = conf.get(CONF_EVL_VERSION)
+    user = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASS)
+    keep_alive = conf.get(CONF_EVL_KEEPALIVE)
+    zone_dump = conf.get(CONF_ZONEDUMP_INTERVAL)
+    zones = conf.get(CONF_ZONES)
+    partitions = conf.get(CONF_PARTITIONS)
+    sync_connect = asyncio.Future(loop=hass.loop)
+
+    controller = EnvisalinkAlarmPanel(
+        host, port, panel_type, version, user, password, zone_dump,
+        keep_alive, hass.loop)
+    hass.data[DATA_EVL] = controller
+
+    @callback
     def login_fail_callback(data):
         """Callback for when the evl rejects our login."""
         _LOGGER.error("The envisalink rejected your credentials.")
-        _connect_status['fail'] = 1
+        sync_connect.set_result(False)
 
+    @callback
     def connection_fail_callback(data):
         """Network failure callback."""
         _LOGGER.error("Could not establish a connection with the envisalink.")
-        _connect_status['fail'] = 1
+        sync_connect.set_result(False)
 
+    @callback
     def connection_success_callback(data):
         """Callback for a successful connection."""
         _LOGGER.info("Established a connection with the envisalink.")
-        _connect_status['success'] = 1
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_envisalink)
+        sync_connect.set_result(True)
 
+    @callback
     def zones_updated_callback(data):
         """Handle zone timer updates."""
         _LOGGER.info("Envisalink sent a zone update event.  Updating zones...")
-        dispatcher.send(signal=SIGNAL_ZONE_UPDATE,
-                        sender=None,
-                        zone=data)
+        hass.bus.async_fire(EVENT_ZONE_UPDATE, {ATTR_ZONE: data})
 
+    @callback
     def alarm_data_updated_callback(data):
         """Handle non-alarm based info updates."""
         _LOGGER.info("Envisalink sent new alarm info. Updating alarms...")
-        dispatcher.send(signal=SIGNAL_KEYPAD_UPDATE,
-                        sender=None,
-                        partition=data)
+        hass.bus.async_fire(EVENT_KEYPAD_UPDATE, {ATTR_PARTITION: data})
 
+    @callback
     def partition_updated_callback(data):
         """Handle partition changes thrown by evl (including alarms)."""
         _LOGGER.info("The envisalink sent a partition update event.")
-        dispatcher.send(signal=SIGNAL_PARTITION_UPDATE,
-                        sender=None,
-                        partition=data)
+        hass.bus.async_fire(EVENT_PARTITION_UPDATE, {ATTR_PARTITION: data})
```

Comment:
(If I remember correctly, @Cinntax initially had it over the bus but I had him change it to dispatcher)